framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,656 @@
+import org.checkerframework.checker.signature.qual.BinaryName;
+import org.checkerframework.checker.signature.qual.DotSeparatedIdentifiers;
+import org.checkerframework.common.wholeprograminference.scenelib.AClassWrapper;
+import org.checkerframework.common.wholeprograminference.scenelib.AFieldWrapper;
+import org.checkerframework.common.wholeprograminference.scenelib.AMethodWrapper;
+import org.checkerframework.common.wholeprograminference.scenelib.ASceneWrapper;
+import org.checkerframework.javacutil.BugInCF;
+import scenelib.annotations.Annotation;
+import scenelib.annotations.el.AMethod;
+import scenelib.annotations.el.AScene;
+import scenelib.annotations.el.ATypeElement;
+import scenelib.annotations.el.AnnotationDef;
+import scenelib.annotations.el.DefCollector;
+import scenelib.annotations.el.DefException;
+import scenelib.annotations.el.InnerTypeLocation;
+import scenelib.annotations.field.AnnotationFieldType;
+import scenelib.annotations.io.IndexFileWriter;
+
+// In this file, "base type" means "type without its package part".
+
+/**
+ * SceneToStubWriter provides static methods that write a {@link AScene} in stub file format, to a
+ * {@link Writer} {@link #write(ASceneWrapper, Writer)} or to a file {@link #write(ASceneWrapper,
+ * Writer)}. This class is the equivalent of {@code IndexFileWriter} from the Annotation File
+ * Utilities, but outputs the results in the stub file format instead of jaif format.
+ *
+ * <p>This writer is used instead of {@code IndexFileWriter} if the {@code -Ainfer=stubs}
+ * command-line argument is present.
+ */
+public final class SceneToStubWriter {
+
+    /** A pattern matching one or more digits. */
+    private static final Pattern digitPattern = Pattern.compile(".*\\$\\d+(\\$.*|$)");
+
+    /** How far to indent when writing members of a stub file. */
+    private static final String INDENT = "    ";
+
+    /**
+     * Writes the annotations in {@code scene} to {@code out} in stub file format.
+     *
+     * @param scene the scene to write out
+     * @param out the Writer to output to
+     */
+    public static void write(ASceneWrapper scene, Writer out) {
+        writeImpl(scene, new PrintWriter(out));
+    }
+
+    /**
+     * Writes the annotations in {@code scene} to the file {@code filename} in stub file format.
+     *
+     * @param scene the scene to write out
+     * @param filename the path of the file to write
+     * @throws IOException if there is trouble writing the file
+     * @see #write(ASceneWrapper, Writer)
+     */
+    public static void write(ASceneWrapper scene, String filename) throws IOException {
+        write(scene, new FileWriter(filename));
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the package.
+     *
+     * @param className the binary name of a class
+     * @return the part of the name referring to the package
+     */
+    // Substrings of binary names are also binary names; the empty string
```

Comment:
I'm just going to remove the signature annotation on the return type here, since there doesn't appear to be one that matches "any valid package name"

Reply:
I think the better design is to use `null` to represent the unnamed package.

Reply:
This still doesn't typecheck without a warning, but I made the change because I agree it's cleaner. The suppression justification/type is more sensical now.

Reply:
That's what I had in mind (independent of the type-checking).  Thanks.